src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs

```diff
@@ -167,25 +168,88 @@ public static LSP.Range TextSpanToRange(TextSpan textSpan, SourceText text)
+        /// </summary>
+        public static async Task<LSP.TextDocumentEdit[]> ChangedDocumentsToTextDocumentEditsAsync<T>(IEnumerable<DocumentId> changedDocuments, Func<DocumentId, T> getNewDocumentFunc,
+                Func<DocumentId, T> getOldDocumentFunc, IDocumentTextDifferencingService? textDiffService, CancellationToken cancellationToken) where T : TextDocument
         {
-            var spanMappingService = document.Services.GetService<ISpanMappingService>();
+            //var uriToTextEdits = new MultiDictionary<Uri, LSP.TextEdit>();
+            using var _ = ArrayBuilder<(Uri Uri, LSP.TextEdit TextEdit)>.GetInstance(out var uriToTextEdits);
 
-            if (spanMappingService == null)
+            foreach (var docId in changedDocuments)
             {
-                return await ConvertTextSpanToLocation(document, textSpan, cancellationToken).ConfigureAwait(false);
+                var newDocument = getNewDocumentFunc(docId);
+                var oldDocument = getOldDocumentFunc(docId);
+
+                var oldText = await oldDocument.GetTextAsync(cancellationToken).ConfigureAwait(false);
+
+                ImmutableArray<TextChange> textChanges;
+
+                // Normal documents have a unique service for calculating minimal text edits. If we used the standard 'GetTextChanges'
+                // method instead, we would get a change that spans the entire document, which we ideally want to avoid.
+                if (newDocument is Document newDoc && oldDocument is Document oldDoc)
+                {
+                    Contract.ThrowIfNull(textDiffService);
+                    textChanges = await textDiffService.GetTextChangesAsync(oldDoc, newDoc, cancellationToken).ConfigureAwait(false);
+                }
+                else
+                {
+                    var newText = await newDocument.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                    textChanges = newText.GetTextChanges(oldText).ToImmutableArray();
+                }
+
+                // Map all the text changes' spans for this document.
+                var mappedResults = await GetMappedSpanResultAsync(oldDocument, textChanges.Select(tc => tc.Span).ToImmutableArray(), cancellationToken).ConfigureAwait(false);
+                if (mappedResults == null)
+                {
+                    // There's no span mapping available, just create text edits from the original text changes.
+                    foreach (var textChange in textChanges)
+                    {
+                        uriToTextEdits.Add((oldDocument.GetURI(), TextChangeToTextEdit(textChange, oldText)));
+                    }
+                }
+                else
+                {
+                    // We have mapping results, so create text edits from the mapped text change spans.
+                    for (var i = 0; i < textChanges.Length; i++)
+                    {
+                        var mappedSpan = mappedResults.Value[i];
```

Comment:
Nit: Could have the contract from here:
https://github.com/dotnet/roslyn/pull/49220/files#diff-c79c7b84c9b0242f40bbdfabae1fd869042e029d50245be3dc18f7ceced25f32R572

Reply:
So the contract is already validated if we get here since it runs in GetMappedSpanResultAsync.  I put it there rather than having all callers assert the contract